src/heliaFetch.ts

```diff
@@ -0,0 +1,144 @@
+import { type UnixFS, unixfs } from '@helia/unixfs'
+import { MemoryBlockstore } from 'blockstore-core'
+import { MemoryDatastore } from 'datastore-core'
+import { type Helia, createHelia } from 'helia'
+import { LRUCache } from 'lru-cache'
+import { CID } from 'multiformats/cid'
+import pTryEach from 'p-try-each'
+
+const ROOT_FILE_PATTERNS = [
+  'index.html',
+  'index.htm',
+  'index.shtml'
+]
+
+const DELEGATED_ROUTING_API = 'https://node3.delegate.ipfs.io/api/v0/name/resolve/'
+
+/**
+ * Fetches files from IPFS or IPNS
+ */
+export class HeliaFetch {
+  private readonly delegatedRoutingApi: string
+  private fs!: UnixFS
+  private readonly rootFilePatterns: string[]
+  public node!: Helia
+  public ready: Promise<void>
+  private readonly ipnsResolutionCache = new LRUCache<string, string>({
+    max: 10000,
+    ttl: 1000 * 60 * 60 * 24
+  })
+
+  constructor (
+    node?: Helia,
+    rootFilePatterns: string[] = ROOT_FILE_PATTERNS,
+    delegatedRoutingApi: string = DELEGATED_ROUTING_API
+  ) {
+    if (node !== undefined) {
+      this.node = node
+    }
+    this.rootFilePatterns = rootFilePatterns
+    this.delegatedRoutingApi = delegatedRoutingApi
+    this.ready = this.init()
+  }
+
+  /**
+   * Initialize the HeliaFetch instance
+   */
+  async init (): Promise<void> {
+    this.node = this.node ?? await createHelia({
+      blockstore: new MemoryBlockstore(),
+      datastore: new MemoryDatastore()
+    })
+    this.fs = unixfs(this.node)
+  }
+
+  /**
+   * Parse a path into its namespace, address, and relative path
+   */
+  public parsePath (path: string): { namespace: string, address: string, relativePath: string } {
+    if (path === undefined) {
+      throw new Error('Path is empty')
+    }
+    const regex = /^\/(?<namespace>ip[fn]s)\/(?<address>[^/$]+)(?<relativePath>[^$]*)/
+    const result = path.match(regex)
+    if (result == null) {
+      throw new Error('Path is not valid, provide path as /ipfs/<cid> or /ipns/<path>')
+    }
+    return result.groups as { namespace: string, address: string, relativePath: string }
```

Comment:
This is returning an error on a valid path:

`http://localhost:8888/ipns/docs.ipfs.tech/how-to/run-ipfs-inside-docker/#set-up`

see https://ipfs.io/ipns/docs.ipfs.tech/how-to/run-ipfs-inside-docker/#set-up

Reply:
though I gave approval, this needs fixed (can be in later PR).